BxlAsLib.txt

```diff
@@ -0,0 +1,377 @@
+
+  - Option 2: Use protobuf-net.
+    - Pros:
+      - Serialization and deserialization come out of box.
+      - Doesn't need a separate proto file.
+    - Cons:
+      - Similar to Google protobuf.
+      - Stability, performance, and encoding size need to be investigated.
+
+  - Option 3 (potentially chosen): Classes have serialization & deserialization methods that mimic
+    protobuf, i.e., there are positional fields.
+    Ex.
+
+    class Data 
+    {       
+        /* TBD: We may want to add positional attrbute for each property, like protobuf-net. */
+
+        public AbsolutePath Path { get; set; }
+        public FileArtifact File { get; set; }
+        public void Serialize(Writer writer) 
+        {
+            if (Path.IsValid) { writer.WritePos(1); writer.WritePath(Path); }
+            if (File.IsValid) { writer.WritePos(2); write.WriteFile(File); }
+        }
+        public static Data Deserialize(Reader reader)
+        {
+            Data data;
+            while (!int pos = reader.ReadPos()) != -1) {
+                switch (pos) {
+                    case 1: { data.Path = reader.ReadPath(); break; }
+                    case 2: { data.File = reader.ReadFile(); break; }
+                }
+            }
+            return data;
+        }
+    }
+
+    - Pros: Can use custom serialization/deserialization.
+    - Cons: 
+      - Manually maintained the position.
+      - An intern working on XLG++ may choose protobuf for versioning.
+
+- Excerpt of ProcessData: basically it reflects Process in BuildXL.Pips
+
+  class ProcessData 
+  {
+      /* TBD: We may want to add positional attrbute for each property, like protobuf-net. */
+
+      public FileArtifact Executable { get; set; }
+      public PipData Arguments { get; set; }
+      public AbsolutePath WorkingDirectory { get; set; }
+      public IEnumerable<EnvironmentVariable> EnvironmentVariables { get; set; }
+      public ProcessOptions ProcessOptions { get; set; }
+      public IEnumerable<FileArtifact> Dependencies { get; set; }
+      public IEnumerable<DirectoryArtifact> DirectoryDependencies { get; set; }
+      public IEnumerable<FileArtifactWithAttributes> FileOutputs { get; set; }
+      public IEnumerable<DirectoryArtifact> DirectoryOutputs { get; set; }
+      ...
+  }       
+
+Graph fragment interactions
+---------------------------
+- We want graph fragments to be able to consume values from other frontends and produce values
+  that can be consumed by other frontends.
+- Even when stitching multiple graphs, we want to estabilish dependency relation between fragments.
+- API:
+  - To expose outputs to other frontends or other fragments:
+
+    interface IGraphFragment
+    {
+        ...
+        /* The implementation may verify that this fragment indeed produces the output file. */
+        AddOutputFileValue(FullSymbol id, FileArtifact outputFile);
+        AddOutputDirectoryValue(FullSymbol id, DirectoryArtifact outputDirectory);

```

Comment:
```suggestion
        Export(FullSymbol id, DirectoryArtifact outputDirectory);
```

Reply:
Per above, Export is fine but we should not use overloads since we want a hihgly compatible api accross versions. It is too easy to introduce a new method overload that will change the behavior when compiling against a new version.